### PR TITLE
Rework of the card database, xml format and oracle parser

### PR DESCRIPTION
**Pull request by ctrlaltca**
## Short roundup of the initial problem
The current carddatabase structure has been designed around a single game.
This has some implications:
 * it can't support other games very well, since only a few fields are available with an hardcoded name;
 * it's not easy to extend, since even adding a single new field requires modifications to hardcoded data structures, the json parser, the xml import/export methods.

## What will change with this Pull Request?
The `CardInfo` class has been modified removing all the game-specific fields (eg. manacost, pt, loyalty, ...) and adding a single "properties" hashmap used to contain these (and more) informations.
A new `CardInfoPerSet` class has been added (with a similar "properties" hashmap) to contain all the card informations that are bound to a specific set (eg: muid, uuid, rarity, collectors number).

A new "version 4" xml format has been introduced, to make better use of these changes, here's an example:
```xml
<card>
    <name>Mournful Kitten</name>
    <text>Every time you stomp on Mournful Kitten, it meows.</text>
    <prop>
        <side>front</side>
        <maintype>Creature</maintype>
        <colors>G</colors>
        <layout>normal</layout>
        <manacost>1G</manacost>
        <cmc>2</cmc>
        <type>Creature</type>
    </prop>
    <set muid="1234" rarity="common" num="85" uuid="c11eb854-4c03-4abd-aaf1-9718608a34b3">KWU</set>
    <set muid="1235" rarity="common" num="18" uuid="c3f048d9-ca13-485d-ad92-de4695b7dc18">KKK</set>
    <tablerow>1</tablerow>
</card>
```
In order to add a new property to a card just add it to the `<prop/>` structure, and the card database loader will pick it up and it will be available through`CardInfo::getProperty("property_name")`.
The "old" cockatrice xml "version 3" loader is still available and working, so this won't piss off users of custom made xmls.

The cardinfo screen has been modified to be able to display all the available card properties, including clickable related cards:
schermata 2019-01-21 alle 09 10 40

The picture loader had been modified to support generic properties, too:
Examples old => new
* !cardid! => !set:muid!
* !uuid!   => !set:uuid!
* !collectornumber! => !set:num!
New examples:
 * !prop:type!
 * !prop:manacost!

A few more related changes:
 * Code in Cockatrice (server updater, release update checker) now uses Qt's builtin json support instead of Qt-Json;
 * Qt-Json is still needed by oracle (because Qt's own json implementation can't handle files as big as mtgjson's AllFiles.json), so it has been moved from `cockatrice/src` to `oracle/src`;
 * Oracle's parser has been rewritten to support the "properties" idea;
 * A card's "colors" information is now saved in a QString, instead of in a QStringList;
 * The card "main type" is now an xml field, and not recalculated every time; the `getMainCardType()` method has been kept in the cockatricexml3 parser to be able to extract the main type of cards in the old format;
 * The "add related card" feature has been reworked to ensure it doesn't pollute the card name in the card datatabase (84faa20b13dd7b843f62f0d81fdc5eb268fa2684);
 * #3504 fixed
 * Added support for transform cards (aka scryfall's "face" property);
 * Started moving Mtg-related property names to a specific file;

### CODE REVIEW

**Comment by tooomm** [comment]
I had some issues when looking at `Manage sets`. It didn't show some long names and set types, after reloading twice and reopening the app it works.
Then I deleted cards.xml and rerun Oracle to see the white spots again...
unknown

There is definitely something weird happening with enabling/disabling sets too.
I disabled all and manually enabled rna, but i had several sets enabled at next start etc.

Version 2.6.2-custom(b43ed50) (2019-01-21)

**Comment by ctrlaltca** [comment]
@tooomm can you please try it again? at least the first issue is solved (blank info on some sets), and i suppose the other one was caused by the same root problem.

**Comment by tooomm** [comment]
Jep, that looks to be solved now! 👍

**Comment by ctrlaltca** [comment]
Rebased on master to fix fedora build

**Comment by ebbit1q** [comment]
Just checked it out, for some reason it doesn't translate the !uuid! properties in the image urls.
```
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to load picture
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Picture not found on disk, trying to download
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to download picture from url: "https://api.scryfall.com/cards/!uuid!?format=image"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]:  Download failed: "Error transferring https://api.scryfall.com/cards/!uuid!?format=image - server replied: Not Found"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Possible picture at  "https://api.scryfall.com/cards/!uuid!?format=image"  could not be loaded
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to load picture
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Picture not found on disk, trying to download
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to download picture from url: "https://api.scryfall.com/cards/multiverse/!cardid!?format=image"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]:  Download failed: "Error transferring https://api.scryfall.com/cards/multiverse/!cardid!?format=image - server replied: Not Found"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Possible picture at  "https://api.scryfall.com/cards/multiverse/!cardid!?format=image"  could not be loaded
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to load picture
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Picture not found on disk, trying to download
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to download picture from url: "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=!cardid!&type=card"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Possible picture at  "http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=!cardid!&type=card"  could not be loaded
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to load picture
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Picture not found on disk, trying to download
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Trying to download picture from url: "http://gatherer.wizards.com/Handlers/Image.ashx?name=Aberrant Researcher&type=card"
PictureLoader: [card:  "Aberrant Researcher"  set:  "SOI" ]: Image successfully downloaded from  "http://gatherer.wizards.com/Handlers/Image.ashx?name=Aberrant Researcher&type=card"
```

File: cockatrice/src/carddatabase.h [hunk 1]
```diff
@@ -112,178 +116,162 @@ class SetList : public QList<CardSetPtr>
...
-    QString cmc;
-    QString cardtype;
-    QString powtough;
+    // The key used to identify this card in the cache
+    QString pixmapCacheKey;
+    // card text
     QString text;
-    QStringList colors;
-
+    // wether this is not a "real" card but a token
```
Comment (ebbit1q) [inline cockatrice/src/carddatabase.h]: whether*

File: cockatrice/src/player.cpp [hunk 2]
```diff
@@ -1843,14 +1830,14 @@ void Player::eventRevealCards(const Event_RevealCards &event)
     }
 
     if (peeking) {
-        for (int i = 0; i < cardList.size(); ++i) {
-            QString cardName = QString::fromStdString(cardList.at(i)->name());
-            CardItem *card = zone->getCard(cardList.at(i)->id(), QString());
+        for (auto i : cardList) {
```
Comment (ebbit1q) [inline cockatrice/src/player.cpp]: should this be an auto &?

**Comment by ZeldaZach** [comment]
We should look at updating the wiki for how to create custom cards with this new knowledge.

cc: @tooomm

**Comment by tooomm** [comment]
I won't be able too look into this soon.
Also, I never worked with custom cards...

**Comment by tooomm** [comment]
>We should look at updating the wiki for how to create custom cards with this new knowledge.

@ctrlaltca can you take care of this, please?

**Comment by Psithief** [comment]
I can't convert the tokens to version 4 without a guide.

**Pull request merged**